### ISSUE TITLE
dont fetch online sourcemaps on runtime

### ISSUE THREAD

**Issue by milahu** [VSCodium/vscodium#1287]
**Describe the bug**

many javascript files in `lib/vscode/resources/` have online sourcemaps

example:

```console
$ tail -n1 /nix/store/0y692yyqc4gdx0i6k22hc7nzswxqqg9p-vscodium-1.71.2.22258/lib/vscode/resources/app/out/bootstrap.js
//# sourceMappingURL=https://ticino.blob.core.windows.net/sourcemaps/cea9d1a4b17966719d8c5f3cdd914480f7cec47c/core/bootstrap.js.map
```

this is unwanted because

1. vscodium should work offline
2. vscodium should not call home to windows.net (telemetry)

**Please confirm that this problem is VSCodium-specific**
- [ ] This bug doesn't happen if I use Microsoft's Visual Studio Code. It only happens in VSCodium.

**Please confirm that the issue/resolution isn't already documented**
- [x] I checked [the Docs page](https://github.com/VSCodium/vscodium/blob/master/DOCS.md) and my issue is not mentioned there.

**Additional context**

on nixos linux ...

> ```find /nix/store/0y692yyqc4gdx0i6k22hc7nzswxqqg9p-vscodium-1.71.2.22258/lib/vscode/resources/ -name '*.js' | xargs tail -n1 -q | sed 's|//# sourceMappingURL=|\n&|g' | grep -E -o 'https://.*?\.js\.map' >/tmp/vscodium-online-sourcemaps.txt```

[vscodium-online-sourcemaps.txt](https://github.com/VSCodium/vscodium/files/9741573/vscodium-online-sourcemaps.txt)

**Comment by daiyam** [comment]
@milahu yes, they are loaded only if you open the Developer Tools in the Help menu.
Should we keep them? I don't know.

**Comment by milahu** [comment]
also some stack traces will (try to) fetch these sourcemaps

sourcemaps are useful for debugging,
so ideally, prefetch the sourcemaps on build, and patch the urls

**Comment by milahu** [comment]
> some stack traces will (try to) fetch these sourcemaps

... to `${workdir}/sourcemaps`, but this fails in my case

debug console:

> ```Could not read source map for file:///nix/store/xzpplq37c76fnjyl3nzgwdpb038yqvhx-vscodium-1.72.0.22279/lib/vscode/resources/app/out/vs/loader.js: ENOENT: no such file or directory, open '/tmp/project/sourcemaps/9f4a41aac91eb148c23b6956be66db333af2d3a2/core/vs/loader.js.map'```

<details>

```
Could not read source map for file:///nix/store/xzpplq37c76fnjyl3nzgwdpb038yqvhx-vscodium-1.72.0.22279/lib/vscode/resources/app/out/vs/loader.js: ENOENT: no such file or directory, open '/tmp/project/sourcemaps/9f4a41aac91eb148c23b6956be66db333af2d3a2/core/vs/loader.js.map'
Could not read source map for file:///nix/store/xzpplq37c76fnjyl3nzgwdpb038yqvhx-vscodium-1.72.0.22279/lib/vscode/resources/app/out/vs/base/common/performance.js: ENOENT: no such file or directory, open '/tmp/project/sourcemaps/9f4a41aac91eb148c23b6956be66db333af2d3a2/core/vs/base/common/performance.js.map'
Could not read source map for file:///nix/store/xzpplq37c76fnjyl3nzgwdpb038yqvhx-vscodium-1.72.0.22279/lib/vscode/resources/app/out/bootstrap-node.js: ENOENT: no such file or directory, open '/tmp/project/sourcemaps/9f4a41aac91eb148c23b6956be66db333af2d3a2/core/bootstrap-node.js.map'
Could not read source map for file:///nix/store/xzpplq37c76fnjyl3nzgwdpb038yqvhx-vscodium-1.72.0.22279/lib/vscode/resources/app/out/bootstrap-amd.js: ENOENT: no such file or directory, open '/tmp/project/sourcemaps/9f4a41aac91eb148c23b6956be66db333af2d3a2/core/bootstrap-amd.js.map'
Could not read source map for file:///nix/store/xzpplq37c76fnjyl3nzgwdpb038yqvhx-vscodium-1.72.0.22279/lib/vscode/resources/app/out/bootstrap.js: ENOENT: no such file or directory, open '/tmp/project/sourcemaps/9f4a41aac91eb148c23b6956be66db333af2d3a2/core/bootstrap.js.map'
Could not read source map for file:///nix/store/xzpplq37c76fnjyl3nzgwdpb038yqvhx-vscodium-1.72.0.22279/lib/vscode/resources/app/out/bootstrap-fork.js: ENOENT: no such file or directory, open '/tmp/project/sourcemaps/9f4a41aac91eb148c23b6956be66db333af2d3a2/core/bootstrap-fork.js.map'
Could not read source map for file:///nix/store/xzpplq37c76fnjyl3nzgwdpb038yqvhx-vscodium-1.72.0.22279/lib/vscode/resources/app/out/vs/workbench/api/node/extensionHostProcess.js: ENOENT: no such file or directory, open '/tmp/project/sourcemaps/9f4a41aac91eb148c23b6956be66db333af2d3a2/core/vs/workbench/api/node/extensionHostProcess.js.map'
Could not read source map for file:///nix/store/xzpplq37c76fnjyl3nzgwdpb038yqvhx-vscodium-1.72.0.22279/lib/vscode/resources/app/out/vs/workbench/api/node/extensionHostProcess.nls.js: ENOENT: no such file or directory, open '/tmp/project/sourcemaps/9f4a41aac91eb148c23b6956be66db333af2d3a2/core/vs/workbench/api/node/extensionHostProcess.nls.js.map'
Could not read source map for file:///nix/store/xzpplq37c76fnjyl3nzgwdpb038yqvhx-vscodium-1.72.0.22279/lib/vscode/resources/app/extensions/debug-auto-launch/dist/extension.js: ENOENT: no such file or directory, open '/tmp/project/sourcemaps/9f4a41aac91eb148c23b6956be66db333af2d3a2/extensions/debug-auto-launch/dist/extension.js.map'
Could not read source map for file:///nix/store/xzpplq37c76fnjyl3nzgwdpb038yqvhx-vscodium-1.72.0.22279/lib/vscode/resources/app/extensions/ipynb/dist/ipynbMain.js: ENOENT: no such file or directory, open '/tmp/project/sourcemaps/9f4a41aac91eb148c23b6956be66db333af2d3a2/extensions/ipynb/dist/ipynbMain.js.map'
Could not read source map for file:///nix/store/xzpplq37c76fnjyl3nzgwdpb038yqvhx-vscodium-1.72.0.22279/lib/vscode/resources/app/extensions/emmet/dist/node/emmetNodeMain.js: ENOENT: no such file or directory, open '/tmp/project/sourcemaps/9f4a41aac91eb148c23b6956be66db333af2d3a2/extensions/emmet/dist/node/emmetNodeMain.js.map'
Could not read source map for file:///nix/store/xzpplq37c76fnjyl3nzgwdpb038yqvhx-vscodium-1.72.0.22279/lib/vscode/resources/app/extensions/merge-conflict/dist/mergeConflictMain.js: ENOENT: no such file or directory, open '/tmp/project/sourcemaps/9f4a41aac91eb148c23b6956be66db333af2d3a2/extensions/merge-conflict/dist/mergeConflictMain.js.map'
```

</details>

... caused by debugging an vscode extension ([vscode-nix-ide](https://github.com/nix-community/vscode-nix-ide))

**Comment by daiyam** [comment]
To be useful, the sourcemaps need to hosted somewhere since MS don't include them since they would increase the install size of the app.

> vscodium should work offline

Since the url needs to be hardcoded, to use a local url would be difficult (for all systems). And the weight of those files can be an issue to include them adhoc.

It only an issue to debug vscode or one of its components.

**Comment by milahu** [comment]
> Since the url needs to be hardcoded, to use a local url would be difficult (for all systems)

`sourceMappingURL` can be a filename, then its relative to dirname

> MS don't include them since they would increase the install size of the app.

maybe we can compress them as .js.map.gz?

but there is another problem ...

> [vscodium-online-sourcemaps.txt](https://github.com/VSCodium/vscodium/files/9741573/vscodium-online-sourcemaps.txt)

all URLs produce HTTP 404, because sourcemaps are only generated in macos builds, and there is no macos build for my vscode versions

* 1.71.2.22258 = cea9d1a4b17966719d8c5f3cdd914480f7cec47c
* 1.72.0.22279 = 9f4a41aac91eb148c23b6956be66db333af2d3a2

https://github.com/microsoft/vscode/issues/39824
https://github.com/Microsoft/vscode/issues/43487

**Comment by daiyam** [comment]
> `sourceMappingURL` can be a filename, then its relative to dirname

I know... but the app is installed at different locations. Unless if we can specify a standard location like the user profile or in the same directory as the extensions, it will be difficult. 
Then, you will need to download a `tar.gz` or `zip` file and uncompress the maps to that location.

**Comment by daiyam** [comment]
Maybe a script that do all of that??? Locate the app, get the version, download the maps and update the urls.... ???

**Comment by daiyam** [comment]
> all URLs produce HTTP 404, because sourcemaps are only generated in macos builds, and there is no macos build for my vscode versions

Yeah, I forgot... it must be linked the BUILD_SOURCEVERSION which is now different between VSCode and VSCodium (due to the release number)
So a solution is needed, either remove them or find a solution to have them locally...

**Comment by GitMensch** [comment]
> So a solution is needed, either remove them or find a solution to have them locally...

Generating them with the CI build and ideally have them uploaded along with changing the URL.
Additional: a script that change the URLs to the home directory and the option to download them packed.

... wishful thinking ...

**Comment by r3m0t** [comment]
> To be useful, the sourcemaps need to hosted somewhere since MS don't include them since they would increase the install size of the app.
> 
> > vscodium should work offline
> 
> Since the url needs to be hardcoded, to use a local url would be difficult (for all systems). And the weight of those files can be an issue to include them adhoc.
> 
> It only an issue to debug vscode or one of its components.

I think this needs to be an option- I can't debug my vscodium without these source maps, so they should be included. For the VSCodium that's downloadable on GitHub, we can assume nobody will use the source map so the sourceMappingURL line can be removed, or the domain name replaced with a `.invalid` domain.

**Comment by daiyam** [comment]
+1

**Comment by milahu** [comment]
> but there is another problem ...
> 
> > [vscodium-online-sourcemaps.txt](https://github.com/VSCodium/vscodium/files/9741573/vscodium-online-sourcemaps.txt)
> 
> all URLs produce HTTP 404, because sourcemaps are only generated in macos builds

or you know... build vscodium from source #887 #420 

one downside is the long build time
this could be reduced by replacing `tsc` with `swc`, see also https://github.com/microsoft/vscode/pull/163660

**Comment by daiyam** [comment]
+1

**Comment by milahu** [comment]
still an issue

**Comment by milahu** [comment]
[still](https://youtube.com/watch?v=u8iMde8kVho)

**Comment by milahu** [comment]
still an issue

**Comment by daiyam** [comment]
I have a plan on how to fix it. I just have to do it...

**Comment by daiyam** [comment]
@milahu Can you try https://github.com/VSCodium/vscodium-insiders/releases/tag/1.105.06780-insider?

FYI, the sourcemaps are stored there: https://github.com/VSCodium/sourcemaps/releases/tag/insider-545721234c97a5f6a5297636e180cba80365eb4a

**Comment by daiyam** [comment]
Wow, 3 years! Damn!

**Comment by Crystal-RainSlide** [comment]
@daiyam I tried the 1.105.06780-insider release, and can confirm that it works fine for not connecting to Microsoft for sourcemap.

**Comment by milahu** [comment]
> the sourcemaps are stored there

that only moves the problem from windows.net to github.com
~~(or did i get this wrong?)~~

```
$ rg -o -I 'sourceMappingURL=.*' | head -n3
sourceMappingURL=https://github.com/VSCodium/sourcemaps/releases/download/insider-545721234c97a5f6a5297636e180cba80365eb4a/core-bootstrap-fork.js.map
sourceMappingURL=https://github.com/VSCodium/sourcemaps/releases/download/insider-545721234c97a5f6a5297636e180cba80365eb4a/core-cli.js.map
sourceMappingURL=https://github.com/VSCodium/sourcemaps/releases/download/insider-545721234c97a5f6a5297636e180cba80365eb4a/core-vs-base-parts-sandbox-electron-browser-preload.js.map
```

expected:

register a custom URL protocol like

```js
import { protocol } from 'electron';

protocol.registerFileProtocol('myapp', (request, callback) => {
  const url = request.url.replace('myapp://', '');
  callback({ path: `${__dirname}/assets/${url}` });
});
```

and use sourcemap URLs like
`myapp://sourcemaps/core/bootstrap.js.map`

... so this works completely offline

**Comment by daiyam** [comment]
> that only moves the problem from windows.net to github.com
> (or did i get this wrong?)

They were broken so not exactly..

I like to solution of custom url with the some custom settings, we can redirect the maps to the online version, local version or simply disable it.

**Comment by milahu** [comment]
bad bot
